# Day 1 - Part C: Data preparation with Numpy and Pandas

Numpy and Pandas are python libraries which automate most of the data manipulation tasks for you. This notebook will provide an overwier of these and guide you though data manipulation, cleaning and preprocessing steps that need to be performed as a part of data science process. 

# 1. Introduction to Numpy
Numpy adds to python support for large, multidimensional array manipulation. It supports operations such as addition, multiplication, inversion and many other.

For more information refer to: https://docs.scipy.org/doc/numpy-dev/user/quickstart.html

## 1.1 Importing Numpy

In [2]:
import numpy as np

In numpy, you can create arrays from python lists.

## 1.2. Basic numpy operations

In [11]:
some_array = np.array([1, 2, 3, 4, 5])
print(some_array)

[1 2 3 4 5]


A single array can contain different objects: strings, floats, integers etc.

In [22]:
some_array = np.array(['blue', 'red', 'orange', 'white'])
print(some_array)

some_array = np.array([1.5, 2, 3.6, 4.1, 5.17])
print(some_array)

['blue' 'red' 'orange' 'white']
[ 1.5   2.    3.6   4.1   5.17]


you can access elements of an array using [ ], similarly as in python's lists.

In [23]:
some_array[1]

2.0

elements in numpy array are mutable, therefore they can be changed as follows:

In [24]:
some_array[3] = 15
print(some_array)

[  1.5    2.     3.6   15.     5.17]


you can assign the whole numpy array to a variable, however be careful when you are doing it as modifying one array will modify the other array as well!

In [33]:
some_array = np.array([1.5, 2, 3.6, 4.1, 5.17])
some_array2 = some_array

print("Before change:")
print("some_array:", some_array)
print("some_array2:", some_array2)

some_array[3] = -6

print("\nAfter change:")
print("some_array:", some_array)
print("some_array2:", some_array2)

Before change:
some_array: [ 1.5   2.    3.6   4.1   5.17]
some_array2: [ 1.5   2.    3.6   4.1   5.17]

After change:
some_array: [ 1.5   2.    3.6  -6.    5.17]
some_array2: [ 1.5   2.    3.6  -6.    5.17]


as you see, elements in **both** arrays were modified even if only one of them was changed! It is because variables *some_array* and *some_array2* are pointers to arrays. We will not discuss pointers here, and you are not required to know how they work, however keep in mind that assigning numpy arrays to variables will "bound" them together and they will be susceptible to changes made on the other copy.

## 1.3 2D arrays
you can create arrays which are 2-dimensional.

In [39]:
np.array( [ [1, 2, 3], [3, 4, 5] ] )

array([[1, 2, 3],
       [3, 4, 5]])

here, you need to pass [ ] to *np.array()* function, and **inside** the brakets further specify the contents of columns. Columns are separacted by a comma. Data in each columns is again defined within [ ] brakets and separated by a comma. You are provided with come examples below, but feel free to experiment with dimensions!

In [44]:
np.array( [ [1, 2], [3, 3], [4, 5] ] )

array([[1, 2],
       [3, 3],
       [4, 5]])

In [45]:
np.array( [ [1, 2, 3, 3, 4, 5] ] )

array([[1, 2, 3, 3, 4, 5]])

In [46]:
np.array( [ [1], [2], [3], [3], [4], [5] ] )

array([[1],
       [2],
       [3],
       [3],
       [4],
       [5]])

## 1.4 Basic operations on arrays

Numpy is very efficient with array operations. Given two arrays A and B you can add them, multiply them, or calculate the inverse. Examples are provided to you below. For more information refer to: https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.matrix.html

In [50]:
A = np.array([ [1, 2], [0, 4]])
B = np.array([ [4, 0], [3, -1]])

print(A)
print()
print(B)

[[1 2]
 [0 4]]

[[ 4  0]
 [ 3 -1]]


In [51]:
A + B

array([[5, 2],
       [3, 3]])

In [52]:
A - B

array([[-3,  2],
       [-3,  5]])

In [53]:
A * B

array([[ 4,  0],
       [ 0, -4]])

To calculate inverse you have to import ** *inv* **

In [57]:
from numpy.linalg import inv

inv(A)

array([[ 1.  , -0.5 ],
       [ 0.  ,  0.25]])

## 1.5 Functions for creating custom arrays
Numpy provides functions that enable you to generate arrays for you. For example, matrices containing only 1's or 0's, random numbers etc. Examples are provided below.

** *np.zeros( (2, 2) )* ** creates 2x2 matrix containing only zeros, ** *np.ones( (2, 3) )* ** creates 2x3 matrix containing only ones, and ** *np.random.rand(2, 3)* ** creates 3x2 matrix containing random numbers between 0 - 1.

You can create 1-dimensional arrays, 2- or even more dimensional matrices.

In [73]:
np.zeros((2, 2))

array([[ 0.,  0.],
       [ 0.,  0.]])

In [75]:
np.ones((2, 3))

array([[ 1.,  1.,  1.],
       [ 1.,  1.,  1.]])

In [76]:
np.random.rand(3,2)

array([[ 0.1046997 ,  0.77646816],
       [ 0.50355231,  0.92834969],
       [ 0.69782807,  0.01370106]])

In [77]:
np.random.rand(3, 2, 2, 4)

array([[[[ 0.07813112,  0.53344132,  0.15166324,  0.71811386],
         [ 0.80316024,  0.77058829,  0.07709759,  0.64650104]],

        [[ 0.63986262,  0.85501332,  0.72759198,  0.25457561],
         [ 0.34507472,  0.96189534,  0.17849875,  0.23220683]]],


       [[[ 0.90199695,  0.82412248,  0.77324594,  0.35296839],
         [ 0.78333253,  0.11753364,  0.63065222,  0.39319407]],

        [[ 0.25404177,  0.09990884,  0.6579875 ,  0.62468771],
         [ 0.8517834 ,  0.31120845,  0.01097131,  0.26747058]]],


       [[[ 0.81845881,  0.15643635,  0.41735095,  0.10469298],
         [ 0.12688916,  0.25467244,  0.69625261,  0.61037712]],

        [[ 0.1611768 ,  0.79555788,  0.22391815,  0.06251261],
         [ 0.2394673 ,  0.86603143,  0.65930057,  0.67977097]]]])

## 1.6 Slicing arrays
Numpy allows to slice the original array and work on the slide instead of the whole dataset

In [94]:
some_array = np.ones((3, 3))

print(some_array)

slice_array = some_array[0:2, 1:3]

print()
print(slice_array)

[[ 1.  1.  1.]
 [ 1.  1.  1.]
 [ 1.  1.  1.]]

[[ 1.  1.]
 [ 1.  1.]]


You can specify the range which columns/rows should be included by using the format *array[ column_start:column_end, row_start:row_end ]*. Rows and columns in numpy (and also in Pandas) start indexing from 0, therefore if you want to include first and second column you need to use 0:2, if second (index 1) and third (index 2) use 1:3 etc.

Again, remember that working on the slice will modify the original array too! Indexing elements is different in the original array and the slice.

In [95]:
print("Before change:")
print("some_array:\n", some_array)
print("some_array2:\n", slice_array)

slice_array[0, 0] = -6

print("\nAfter change:")
print("some_array:\n", some_array)
print("some_array2:\n", slice_array)

Before change:
some_array:
 [[ 1.  1.  1.]
 [ 1.  1.  1.]
 [ 1.  1.  1.]]
some_array2:
 [[ 1.  1.]
 [ 1.  1.]]

After change:
some_array:
 [[ 1. -6.  1.]
 [ 1.  1.  1.]
 [ 1.  1.  1.]]
some_array2:
 [[-6.  1.]
 [ 1.  1.]]


# 2. Introduction to Pandas

Pandas, similarly to numpy, provides support with large, multidimensional arrays. Additionally it is equipped with many more functions. The following section provides some key information about the use of the library in data manipulation. You are free to read more here: http://pandas.pydata.org/pandas-docs/stable/tutorials.html

## 2.1 Importing Pandas

In [3]:
import pandas as pd

## 2.2 Importing data from files
The ** *read_csv()* ** function enables to read CSV files, however pandas enables to read other formats e.g. excel files. Below, is an exemplary code snipper which reads the file *data.csv* and loads it into pandas DataFrame.

Useful arguments of the function:
- ** *sep* ** specifies the separator with which cells are separated with, 
- *** encoding* ** enables to change the file encoding. The default encoding is *utf-8*, however you might try *ISO-8859-1* if utf-8 does not work.

Run the code below to examine first rows of the dataset with ** *dataFrame.head()* ** function. Try to change the encoding to *utf-8* and see what happens!

In [4]:
dataFrame = pd.read_csv('./datasets/ecommerce_data/e_commerce.csv', sep=',', encoding='ISO-8859-1')
dataFrame.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 2.3 Creating DataFrame from a python dictionary
Alternatively, you can create Pandas DataFrame from a dictionay

In [97]:
data = [{'color_name': 'black', 'R': 0, 'G': 0, 'B': 0}, {'color_name': 'white', 'R': 255, 'G': 255, 'B': 255}, 
        {'color_name': 'red', 'R': 255, 'G': 0, 'B': 0}, {'color_name': 'blue', 'R': 0, 'G': 0, 'B': 255},
        {'color_name': 'green', 'R': 0, 'G': 255, 'B': 0}]

dataFrame = pd.DataFrame(data)

## 2.4 Inspecting the dataset

To inspect the dataset use ** *head()* ** function. This will show you first 5 rows of the dataset.

In [98]:
dataFrame.head()

B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

Hoever, it will work only when ** *head()* ** is executed at the end of the code block. If you want to print your dataframe somewhere in the middle use:

In [99]:
print(dataFrame.head())

     B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green


## 2.5 Selecting, modifying and removing rows/columns

### Selecting a column
A good thing about Pandas is that you can refer to specific columns simply by using its name:

In [101]:
dataFrame['B']

0      0
1    255
2      0
3    255
4      0
Name: B, dtype: int64

In [102]:
dataFrame['color_name']

0    black
1    white
2      red
3     blue
4    green
Name: color_name, dtype: object

### Selecting rows
To select range of rows use the code below. DataFrame[2:4] selects all columns for rows from 2 to 3 (4 is not included)

In [103]:
dataFrame[2:4]

B  G    R color_name
2    0  0  255        red
3  255  0    0       blue

If you want to select specific row indexes use ** *iloc* **:

In [104]:
dataFrame.iloc[[0,2]]

B  G    R color_name
0  0  0    0      black
2  0  0  255        red

You can also select first e.g. 2 rows of the dataset, like below:

In [105]:
dataFrame[:2]

B    G    R color_name
0    0    0    0      black
1  255  255  255      white

Or last rows

In [106]:
dataFrame[-2:]

B    G  R color_name
3  255    0  0       blue
4    0  255  0      green

### Selecting multiple columns
You can select multiple columns by passing it as an array of column names, just like below:

In [107]:
column_slice = dataFrame[['B','color_name']]
column_slice.head()

B color_name
0    0      black
1  255      white
2    0        red
3  255       blue
4    0      green

You can find more on indexing here: http://pandas.pydata.org/pandas-docs/stable/indexing.html

### Column names and indices
You can return column names by using:

In [108]:
dataFrame.columns

Index(['B', 'G', 'R', 'color_name'], dtype='object')

You can see the row numbers/names by using:

In [109]:
dataFrame.index

RangeIndex(start=0, stop=5, step=1)

### Conditional selection
You can apply conditional selection by using ** dataFrame[*column_name*] *condition* *value* ** as specified below. The result of such expression is a True/False data frame which can be used to display rows which fulfil the condition. For example, the code below selects all rows that have *R* higher than 100.

In [110]:
select_greater_than_100 = (dataFrame['R'] > 100)
dataFrame[select_greater_than_100]

B    G    R color_name
1  255  255  255      white
2    0    0  255        red

### Creating new columns
You can create a new column just by assigning values to non-existent column name. You can use completely new values, or create column using values from another one, as shown below: 

In [111]:
dataFrame['color_number'] = 65536 * dataFrame['R'] + 256 * dataFrame['G'] + dataFrame['B']
dataFrame.head()

B    G    R color_name  color_number
0    0    0    0      black             0
1  255  255  255      white      16777215
2    0    0  255        red      16711680
3  255    0    0       blue           255
4    0  255    0      green         65280

### Removing columns
You can remove the column by using ** *del* **:

In [112]:
del dataFrame['color_number']
dataFrame.head()

B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

Or simply pop it from the dataset. This operation will remove the column and return its values creating new DataFrame

In [113]:
new_dataFrame = dataFrame.pop('color_name')
print(new_dataFrame)
print()
print(dataFrame)

0    black
1    white
2      red
3     blue
4    green
Name: color_name, dtype: object

     B    G    R
0    0    0    0
1  255  255  255
2    0    0  255
3  255    0    0
4    0  255    0


and now adding it back to the dataset

In [114]:
dataFrame['color_name'] = new_dataFrame
dataFrame.head()

B    G    R color_name
0    0    0    0      black
1  255  255  255      white
2    0    0  255        red
3  255    0    0       blue
4    0  255    0      green

# 3. Data Science Process

The usual data science process contains the following steps:
1. Initial inspection of the dataset
2. Data preparation (cleaning, feature selection, normalisation and splitting dataset into train/cross-validation/test)
4. Building the model
5. Evaluating the model performance

Before the machine learning can be applied to the dataset, a significant effort needs to be put into preparing the dataset for the model. Poorly performed data preparation might result in poor performance or unreliable results. This jupyter notebook will cover initial inspection of the dataset, data cleaning, normalisation and splitting dataset into train/cross-validation/test. Feature selection will be discussed in more details in notebooks which cover machine learning models.

We will do so on the exemplary e-commerce dataset.

## 3.1 Data cleaning and initial inspection

In [89]:
dataFrame = pd.read_csv('./datasets/ecommerce_data/e_commerce.csv', sep=',', encoding='ISO-8859-1')
dataFrame.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

We can see that the dataset includes 8 columns: *Invoice No.*, *StockCode*, *Description*, *Quantity*, *InvoiceDate*, *UnitPrice*, *CustomerID*, and *Country*.

Numerical variables include: *InvoiceNo*, *Quantity*, *UnitPrice*, and *CustomerID*.

Categorical variables include: *StockCode*, *Description*, and *Country*.

There is only one datetime type variable: *InvoiceDate*.

### 3.1.1 Replacing missing values
We will check if there are any variables that are missing.

In [75]:
dataFrame.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
dtype: bool

We can see that there are *Description* and *CustomerID* features which are empty. The decision what to do with these depends on the circumstances. The most popular options are:

- remove missing values
- replace them with mean/most common category
- replace with the value of the previous row (forward-fill)
- replace with the value of the next row (backward-fill)

#### Removing missing values
Removing the missing values is done using ** *dropna()* ** function. It is important to assign reduced DataFrame to a variable, otherwise changes will not have any effect. Example of such behaviour is presented below.


In [76]:
print(dataFrame.shape)

dataFrame.dropna()

print(dataFrame.shape)

(238735, 8)
(238735, 8)


As you can see, the size of the DataFrame did not change, which means that the empty rows were **NOT** removed. Below is the corrected version of the code. The dataset has been reduced by more than 66 000 samples.

In [90]:
print(dataFrame.shape)

dataFrame = dataFrame.dropna()

print(dataFrame.shape)

(238735, 8)
(171810, 8)


#### Replacing missing values with the mean
First we will set *Quantity* in the row with index 1 to null, and then observe how it has been replaced.

In [78]:
dataFrame.loc[1, 'Quantity'] = np.nan
dataFrame.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER       6.0   
1    536365     71053                  WHITE METAL LANTERN       NaN   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER       8.0   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

Replacing missing values in column *Quantity* with mean value

In [79]:
dataFrame['Quantity'] = (dataFrame['Quantity'].replace(np.nan, dataFrame['Quantity'].mean()))
dataFrame.head()

InvoiceNo StockCode                          Description   Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER   6.000000   
1    536365     71053                  WHITE METAL LANTERN  12.254579   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER   8.000000   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE   6.000000   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.   6.000000   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

#### Replacing missing values with previous row (forward-fill)

First, setting null values in *StockCode* and *InvoiceDate* in the row with index 7. 

In [36]:
dataFrame.loc[7, 'StockCode'] = np.nan
dataFrame.loc[7, 'InvoiceDate'] = np.nan
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366       NaN               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7             NaN       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

Replacing missing values with the previous value (forward fill). Please note that this method will replace missing values in all columns, with values respective to that column; and might find very useful when replacing missing values in time-series.

In [37]:
dataFrame = dataFrame.fillna(method='ffill')
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366     21730               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:26       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

#### Replacing missing values with next row (backward-fill)

In [38]:
dataFrame.loc[7, 'StockCode'] = np.nan
dataFrame.loc[7, 'InvoiceDate'] = np.nan
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366       NaN               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7             NaN       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

In [39]:
dataFrame = dataFrame.fillna(method='bfill')
dataFrame[3:10]

InvoiceNo StockCode                          Description  Quantity  \
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE       6.0   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.       6.0   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES       2.0   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER       6.0   
7    536366     22632               HAND WARMER UNION JACK       6.0   
8    536366     22632            HAND WARMER RED POLKA DOT       6.0   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT      32.0   

      InvoiceDate  UnitPrice  CustomerID         Country  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  
5  12/1/2010 8:26       7.65     17850.0  United Kingdom  
6  12/1/2010 8:26       4.25     17850.0  United Kingdom  
7  12/1/2010 8:28       1.85     17850.0  United Kingdom  
8  12/1/2010 8:28       1.85     17850.0  United Kingdom  
9  12/1/2010 8:34       1.69     13047.0  United Kingdom

### 3.1.2 Initial inspection and data cleaning

Now we can investigate how the numerical dataset looks like by using ** *describe()* ** function.

In [40]:
dataFrame.describe()

Quantity      UnitPrice     CustomerID
count  171810.000000  171810.000000  171810.000000
mean       12.254579       3.798716   15274.144730
std       258.624198     101.355052    1724.536126
min    -74215.000000       0.000000   12346.000000
25%         2.000000       1.250000   13838.000000
50%         5.000000       1.950000   15130.000000
75%        12.000000       3.750000   16814.000000
max     74215.000000   38970.000000   18287.000000

It can be observed that the *Quantity* feature can have negative values, which is not what we would expect. These values can be removed from the dataset by applying conditional selection discussed before.

First, we see how many of these have non-positive *Quantity* values.

In [41]:
len(dataFrame[dataFrame['Quantity'] <= 0])

4098

Now, we will be selecting these rows using conditional selection.

In [91]:
print(dataFrame.shape)

indexes_with_negative_height = dataFrame[dataFrame['Quantity'] < 0].index
dataFrame = dataFrame.drop(indexes_with_negative_height)

print(dataFrame.shape)

(171810, 8)
(167712, 8)


We can observe that the dataset has been reduces exactly by 4098 rows.

Interestingly, there are 13 rows which have unit price equal to 0. This raises a warning flag since it is not often for an item to be for free. We will be displaying these rows.

In [43]:
len(dataFrame[dataFrame['UnitPrice'] == 0])

13

In [44]:
dataFrame[dataFrame['UnitPrice'] == 0]

InvoiceNo StockCode                          Description  Quantity  \
9302      537197     22841         ROUND CAKE TIN VINTAGE GREEN       1.0   
33576     539263     22580         ADVENT CALENDAR GINGHAM SACK       4.0   
40089     539722     22423             REGENCY CAKESTAND 3 TIER      10.0   
47068     540372     22090              PAPER BUNTING RETROSPOT      24.0   
47070     540372     22553               PLASTERS IN TIN SKULLS      24.0   
56674     541109     22168        ORGANISER WOOD ANTIQUE WHITE        1.0   
86789     543599    84535B         FAIRY CAKES NOTEBOOK A6 SIZE      16.0   
130188    547417     22062  CERAMIC BOWL WITH LOVE HEART DESIGN      36.0   
139453    548318     22055   MINI CAKE STAND  HANGING STRAWBERY       5.0   
145208    548871     22162          HEART GARLAND RUSTIC PADDED       2.0   
157042    550188     22636   CHILDS BREAKFAST SET CIRCUS PARADE       1.0   
187613    553000     47566                        PARTY BUNTING       4.0   
198383    554037     22619            SET OF 6 SOLDIER SKITTLES      80.0   

             InvoiceDate  UnitPrice  CustomerID         Country  
9302     12/5/2010 14:02        0.0     12647.0         Germany  
33576   12/16/2010 14:36        0.0     16560.0  United Kingdom  
40089   12/21/2010 13:45        0.0     14911.0            EIRE  
47068     1/6/2011 16:41        0.0     13081.0  United Kingdom  
47070     1/6/2011 16:41        0.0     13081.0  United Kingdom  
56674    1/13/2011 15:10        0.0     15107.0  United Kingdom  
86789    2/10/2011 13:08        0.0     17560.0  United Kingdom  
130188   3/23/2011 10:25        0.0     13239.0  United Kingdom  
139453   3/30/2011 12:45        0.0     13113.0  United Kingdom  
145208    4/4/2011 14:42        0.0     14410.0  United Kingdom  
157042   4/14/2011 18:57        0.0     12457.0     Switzerland  
187613   5/12/2011 15:21        0.0     17667.0  United Kingdom  
198383   5/20/2011 14:13        0.0     12415.0       Australia

However unusual, these items might still be valid entries because they might have been sold as an addition to other item, or bought with a voucher, therefore they will be kept in the dataset.

You can further inspect the dataset by calling  functions e.g. min(), max(), mean(), etc. explicitly on the dataFrame

In [45]:
dataFrame.max()

InvoiceNo                                557955
StockCode                                  POST
Description    ZINC WIRE SWEETHEART LETTER TRAY
Quantity                                  74215
InvoiceDate                       6/9/2011 9:54
UnitPrice                               8142.75
CustomerID                                18287
Country                             Unspecified
dtype: object

Note that for the strings the minimum and maximum values are evaluated based on the characters included in the string. For example, a letter "A" will be considered as "smaller" than letter "Z". Moreoever, string digits e.g. "8" are always considered to be smaller than letters.

In [46]:
dataFrame.min()

InvoiceNo                              536365
StockCode                               10002
Description     4 PURPLE FLOCK DINNER CANDLES
Quantity                                    1
InvoiceDate                   1/10/2011 10:32
UnitPrice                                   0
CustomerID                              12346
Country                             Australia
dtype: object

Similarly, you can apply functions ** *mode()* ** or ** *mean()* **.

More functions can be found here: https://pandas.pydata.org/pandas-docs/stable/api.html#api-dataframe-stats

You can check out the correlation coefficients between columns of numerical values by using *corr()* function. In the next notebook on data visualisation you will learn how to create a correlation heatmap.

In [47]:
dataFrame.corr()

Quantity  UnitPrice  CustomerID
Quantity    1.000000  -0.004563   -0.010708
UnitPrice  -0.004563   1.000000   -0.007881
CustomerID -0.010708  -0.007881    1.000000

### Invalid categorical values

If you want to inspect if the categorical column contains invalid entries you can use *value_counts()* function, as below.

In [48]:
dataFrame['Country'].value_counts()

United Kingdom          149183
Germany                   3949
France                    3531
EIRE                      2609
Netherlands               1139
Spain                     1131
Belgium                    914
Switzerland                679
Australia                  630
Portugal                   610
Norway                     368
Channel Islands            363
Cyprus                     350
Finland                    307
Italy                      299
Japan                      230
Sweden                     195
Denmark                    184
Poland                     180
Austria                    124
Singapore                  113
Iceland                    102
Greece                      85
Unspecified                 72
Canada                      68
Lebanon                     45
Malta                       45
Lithuania                   35
Brazil                      32
European Community          31
United Arab Emirates        30
USA                         22
Bahrain 

The same applies for numerical values

Most of the values seem to be valid entries, however some of the countries are categorized as "Unspecified". It is our decision whether we want to remove rows which have this value or not, and in this case we will do this be executing the following code:

In [92]:
print(dataFrame.shape)

dataFrame = dataFrame.drop(dataFrame[dataFrame['Country'] == 'Unspecified'].index)

print(dataFrame.shape)

(167712, 8)
(167640, 8)


We have removed exactly 72 rows, as specified for 'Unspecified' by value_counts().

## 3.2 Encoding categorical variables

Most machine learning methods do not directly work with categories, there is a need to transform these into numerical values. There are two options to do it:

- encoding each category by number e.g. Monday - 1, Tuesday - 2 etc.
- creating dummy columns which encode the category. For example for day of the week, we would create 7 columns, each corresponsing to one day of the week.

To demonstrate both approaches, we will switch to the color dataset used at the beginning of the notebook.

In [54]:
data = [{'color_name': 'black', 'R': 0, 'G': 0, 'B': 0}, {'color_name': 'white', 'R': 255, 'G': 255, 'B': 255}, 
        {'color_name': 'red', 'R': 255, 'G': 0, 'B': 0}, {'color_name': 'blue', 'R': 0, 'G': 0, 'B': 255},
        {'color_name': 'green', 'R': 0, 'G': 255, 'B': 0}, {'color_name': 'green', 'R': 0, 'G': 255, 'B': 0}]

new_dataFrame = pd.DataFrame(data)
new_dataFrame['color_name2'] = new_dataFrame['color_name']
new_dataFrame

B    G    R color_name color_name2
0    0    0    0      black       black
1  255  255  255      white       white
2    0    0  255        red         red
3  255    0    0       blue        blue
4    0  255    0      green       green
5    0  255    0      green       green

### 3.2.1 Numerical encoding of categorical variables

First, will transform the type of the column to *category*, so that pandas map each category with the unique code.

In [58]:
new_dataFrame['color_name'] = new_dataFrame['color_name'].astype('category')

Next, we will update the column with corresponding category codes

In [59]:
new_dataFrame['color_name'] = new_dataFrame['color_name'].cat.codes
new_dataFrame

B    G    R  color_name color_name2
0    0    0    0           0       black
1  255  255  255           4       white
2    0    0  255           3         red
3  255    0    0           1        blue
4    0  255    0           2       green
5    0  255    0           2       green

We can see that each color has now each category assigned: *black* - 0, *white* - 4, *red* - 3, etc.

### 3.2.2 Creating dummy columns (also called as One Hot Encoding)

Execute ** *pd.get_dummies()* ** function to expand a specific column to dummy columns. *columns* parameter specifies the names of columns that will be expanded using one hot encoding.

In [60]:
new_dataFrame = pd.get_dummies(new_dataFrame, columns=['color_name2'])

In [61]:
new_dataFrame

B    G    R  color_name  color_name2_black  color_name2_blue  \
0    0    0    0           0                  1                 0   
1  255  255  255           4                  0                 0   
2    0    0  255           3                  0                 0   
3  255    0    0           1                  0                 1   
4    0  255    0           2                  0                 0   
5    0  255    0           2                  0                 0   

   color_name2_green  color_name2_red  color_name2_white  
0                  0                0                  0  
1                  0                0                  1  
2                  0                1                  0  
3                  0                0                  0  
4                  1                0                  0  
5                  1                0                  0

Now we can see that the *color_name2* column has been expanded to 5 binary columns.

## 3.3 Unbalanced datasets

Datasets often might contain unbalanced categorical variables; that means that some of these values appear more frequenty than the other. When the unbalanced feature is the target i.e. the feature that we want to predict using machine learning models, it poses a problem because our model might "learn" to predict the frequent class only and it might be right for the most of the time. 

For example, let's assume that a manufacturing company owns a machine and want to predict whenther it will break or not. If the machine works for 99% of the time and is broken only 1% of the time, predicting that it will be functional all the time will give us 99% accuracy, however will completely neglect the cases when it might be broken. 

We certainly don't want our model to focus on the features of the most frequent class. There are certain techniques to help us to improve machine learning model performance in datasets which are unbalanced. Examples of these methods are: downsampling and upsampling. Downsampling reduce the size of our original dataset removing instances of the majority class to match the minority class. Upsampling adds new data points which are similar to these of the minority class. 

Downsampling and upsampling will be presented below, however before we will slighlty modify the e-commerce dataset.

Let's assume that we are interested only in transactions which happened in United Kingdom and France. Below is the code which will select all the rows which have 'United Kingdom' value in column *Country*, and these with 'France' value in the same column.

Both selections are then concatenated togehter using ** *pd.concat()* ** pandas function. What concatenate is doing is adding one data frame below the other; this implies that both data frames need tohave the same columns. More about concatenation (and other merging functions) can be found here: https://pandas.pydata.org/pandas-docs/stable/merging.html

In [93]:
UKData = dataFrame.loc[dataFrame['Country'] == 'United Kingdom']
FranceData = dataFrame.loc[dataFrame['Country'] == 'France']

dataFrame = pd.concat([UKData, FranceData])
dataFrame

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5         536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6         536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7         536366     22633               HAND WARMER UNION JACK         6   
8         536366     22632            HAND WARMER RED POLKA DOT         6   
9         536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   
10        536367     22745           POPPY'S PLAYHOUSE BEDROOM          6   
11        536367     22748            POPPY'S PLAYHOUSE KITCHEN         6   
12        536367     22749    FELTCRAFT PRINCESS CHARLOTTE DOLL         8   
13        536367     22310              IVORY KNITTED MUG COSY          6   
14        536367     84969   BOX OF 6 ASSORTED COLOUR TEASPOONS         6   
15        536367     22623        BOX OF VINTAGE JIGSAW BLOCKS          3   
16        536367     22622       BOX OF VINTAGE ALPHABET BLOCKS         2   
17        536367     21754             HOME BUILDING BLOCK WORD         3   
18        536367     21755             LOVE BUILDING BLOCK WORD         3   
19        536367     21777          RECIPE BOX WITH METAL HEART         4   
20        536367     48187                  DOORMAT NEW ENGLAND         4   
21        536368     22960             JAM MAKING SET WITH JARS         6   
22        536368     22913          RED COAT RACK PARIS FASHION         3   
23        536368     22912       YELLOW COAT RACK PARIS FASHION         3   
24        536368     22914         BLUE COAT RACK PARIS FASHION         3   
25        536369     21756             BATH BUILDING BLOCK WORD         3   
46        536371     22086      PAPER CHAIN KIT 50'S CHRISTMAS         80   
47        536372     22632            HAND WARMER RED POLKA DOT         6   
48        536372     22633               HAND WARMER UNION JACK         6   
49        536373    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
...          ...       ...                                  ...       ...   
235974    557669     22629                  SPACEBOY LUNCH BOX         12   
235975    557669     22631             CIRCUS PARADE LUNCH BOX         12   
235976    557669     22319    HAIRCLIPS FORTIES FABRIC ASSORTED        12   
235977    557669     23076           ICE CREAM SUNDAE LIP GLOSS        24   
235978    557669     22328   ROUND SNACK BOXES SET OF 4 FRUITS          6   
235979    557669     22326  ROUND SNACK BOXES SET OF4 WOODLAND          6   
235980    557669     21881                       CUTE CATS TAPE        12   
235981    557669     21883                     STARS GIFT TAPE         12   
235982    557669     16237                 SLEEPING CAT ERASERS        30   
235983    557669     22492              MINI PAINT SET VINTAGE         36   
235984    557669     22727            ALARM CLOCK BAKELIKE RED          4   
235985    557669     22726           ALARM CLOCK BAKELIKE GREEN         4   
235986    557669     22728            ALARM CLOCK BAKELIKE PINK         4   
235987    557669     22398           MAGNETS PACK OF 4 SWALLOWS        12   
235988    557669     23158          SET OF 5 LUCKY CAT MAGNETS         12   
235989    557669      POST                              POSTAGE         2   
236116    557734     20974          12 PENCILS SMALL TUBE SKULL        48   
236117    557734     23190  BUNDLE OF 3 SCHOOL EXERCISE BOOKS          12   
236118    557734     22149           FELTCRAFT 6 FLOWER FRIENDS         6   
236119    557734     23126            FELTCRAFT GIRL AMELIE KIT         4   
236

Now we can confirm that our dataset contain only entries related to United Kingdom and France. Moreover, we can observe that there are many more entries related to UK than to France. This means that this dataset is *unbalanced*, therefore we will be applying downsampling and upsampling.

In [82]:
dataFrame['Country'].value_counts()

United Kingdom    149183
France              3531
Name: Country, dtype: int64

### 3.3.1 Downsampling

Downsampling is removing the instances of the majority class so that it matches the number of minority class instances. We will do it using sklearn library ** *resample* **. First, we will create slices of the original dataset, where the first slice contains the majority class ("United Kingdom"), and the second slice contains the minorty class ("France").

In [94]:
from sklearn.utils import resample

df_majority = dataFrame[dataFrame['Country']=='United Kingdom']
df_minority = dataFrame[dataFrame['Country']=='France']
minority_len = (len(df_minority))

Now, we will run the resample method with parameter *replace* equal to False, and n_samples equal to the number of instances in the minority class. The reduced majority class will be concatenated with the minority class, and we can observe that there is now exactly the same number of instances for majority and minority classes.

In [ ]:
df_majority_downsampled = resample(df_majority,replace=False,n_samples=minority_len,random_state=123)
ddown = pd.concat([df_minority, df_majority_downsampled])
ddown['Country'].value_counts()

### 3.3.2 Upsampling

Upsampling will create copies of dataset instances to match required number of samples. Below, we count number of instances in the majority class, and create additional instances in the minority class to match the majority class.

We can now see that there is the same number of instances for both classes, and it is equal to the initial size of the majority class.

In [95]:
majority_len = (len(df_majority))

df_minority_upsampled = resample(df_minority,replace=True,n_samples=majority_len,random_state=123)
dup = pd.concat([df_majority, df_minority_upsampled])
dup['Country'].value_counts()

United Kingdom    149183
France            149183
Name: Country, dtype: int64

## 3.4 Normalization

Most of the machine learning algorithms require the dataset they are trained on to be normalized. Below is the example of normalization of the e-commerce dataset.

First, we will be encoding categorical variables. We will use category codes for *StockCode* and *Description*, and One Hot Encoding for *Contry*.

In [96]:
dataFrame.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [97]:
dataFrame['StockCode'] = dataFrame['StockCode'].astype('category')
dataFrame['StockCode'] = dataFrame['StockCode'].cat.codes

dataFrame['Description'] = dataFrame['Description'].astype('category')
dataFrame['Description'] = dataFrame['Description'].cat.codes

dataFrame = pd.get_dummies(dataFrame, columns=['Country'])
dataFrame.head()

InvoiceNo  StockCode  Description  Quantity     InvoiceDate  UnitPrice  \
0    536365       2886         3173         6  12/1/2010 8:26       2.55   
1    536365       2329         3181         6  12/1/2010 8:26       3.39   
2    536365       2514          774         8  12/1/2010 8:26       2.75   
3    536365       2468         1588         6  12/1/2010 8:26       3.39   
4    536365       2467         2416         6  12/1/2010 8:26       3.39   

   CustomerID  Country_France  Country_United Kingdom  
0     17850.0               0                       1  
1     17850.0               0                       1  
2     17850.0               0                       1  
3     17850.0               0                       1  
4     17850.0               0                       1

The *InvoiceDate* field would need to be translated into the timestamp in order to be correcrtly normalized, however as it is not the scope of this notebook we will remove it from our data frame. If you are interested, you can read more about timestamps here: https://pandas.pydata.org/pandas-docs/stable/timeseries.html

In [98]:
del dataFrame['InvoiceDate']

We will use ** *preprocessing* ** package from sklearn to normalize our dataset. We have to normalize column by column. Otherwise the normalization will take place across different column.

The preprocessing package requires to pass the data in a form of a vector, therefore we will be first reshaping pandas data frame to a vector, and then reshaping it again as a column and updating the data frame.

In [99]:
from sklearn import preprocessing

for column in dataFrame.columns:
    vector = dataFrame[column].values.reshape(1, len(dataFrame[column]))
    normalized_vector = preprocessing.normalize(vector, norm="l2")
    dataFrame[column] = normalized_vector.reshape(len(dataFrame[column]), 1)
dataFrame.head()

InvoiceNo  StockCode  Description  Quantity  UnitPrice  CustomerID  \
0   0.002508   0.004574     0.004132  0.000078   0.000290    0.002934   
1   0.002508   0.003691     0.004142  0.000078   0.000385    0.002934   
2   0.002508   0.003984     0.001008  0.000105   0.000312    0.002934   
3   0.002508   0.003911     0.002068  0.000078   0.000385    0.002934   
4   0.002508   0.003910     0.003146  0.000078   0.000385    0.002934   

   Country_France  Country_United Kingdom  
0             0.0                0.002589  
1             0.0                0.002589  
2             0.0                0.002589  
3             0.0                0.002589  
4             0.0                0.002589